ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp

```diff
@@ -1,17 +1,254 @@
+{
+  draw_primitive_begin(kind);
+}
+
+void d3d_primitive_begin_texture(int kind, int texId)
+{
+  draw_primitive_begin_texture(kind, texId);
+}
+
+void d3d_primitive_end()
+{
+  draw_primitive_end();
+}
+
+void d3d_vertex(gs_scalar x, gs_scalar y, gs_scalar z)
+{
+  d3d_model_vertex_color(draw_get_batch_stream(), x, y, z, draw_get_color(), draw_get_alpha());
+}
+
+void d3d_vertex_color(gs_scalar x, gs_scalar y, gs_scalar z, int color, double alpha)
+{
+  d3d_model_vertex_color(draw_get_batch_stream(), x, y, z, color, alpha);
+}
+
+void d3d_vertex_texture(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar tx, gs_scalar ty)
+{
+  d3d_model_vertex_texture_color(draw_get_batch_stream(), x, y, z, tx, ty, draw_get_color(), draw_get_alpha());
+}
+
+void d3d_vertex_texture_color(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar tx, gs_scalar ty, int color, double alpha)
+{
+  d3d_model_vertex_texture_color(draw_get_batch_stream(), x, y, z, tx, ty, color, alpha);
+}
+
+void d3d_vertex_normal(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar nx, gs_scalar ny, gs_scalar nz)
+{
+  d3d_model_vertex_normal_color(draw_get_batch_stream(), x, y, z, nx, ny, nz, draw_get_color(), draw_get_alpha());
+}
+
+void d3d_vertex_normal_color(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar nx, gs_scalar ny, gs_scalar nz, int color, double alpha)
+{
+  d3d_model_vertex_normal_color(draw_get_batch_stream(), x, y, z, nx, ny, nz, color, alpha);
+}
+
+void d3d_vertex_normal_texture(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar nx, gs_scalar ny, gs_scalar nz, gs_scalar tx, gs_scalar ty)
+{
+  d3d_model_vertex_normal_texture_color(draw_get_batch_stream(), x, y, z, nx, ny, nz, tx, ty, draw_get_color(), draw_get_alpha());
+}
+
+void d3d_vertex_normal_texture_color(gs_scalar x, gs_scalar y, gs_scalar z, gs_scalar nx, gs_scalar ny, gs_scalar nz, gs_scalar tx, gs_scalar ty, int color, double alpha)
+{
+  d3d_model_vertex_normal_texture_color(draw_get_batch_stream(), x, y, z, nx, ny, nz, tx, ty, color, alpha);
+}
+
+void d3d_draw_floor(gs_scalar x1, gs_scalar y1, gs_scalar z1, gs_scalar x2, gs_scalar y2, gs_scalar z2, int texId, gs_scalar hrep, gs_scalar vrep)
+{
+  draw_batch_begin_deferred(texId);
+  d3d_model_floor(draw_get_batch_stream(), x1, y1, z1, x2, y2, z2, hrep, vrep);
+  draw_batch_flush(batch_flush_immediate);
+}
+
+void d3d_draw_wall(gs_scalar x1, gs_scalar y1, gs_scalar z1, gs_scalar x2, gs_scalar y2, gs_scalar z2, int texId, gs_scalar hrep, gs_scalar vrep)
+{
+  draw_batch_begin_deferred(texId);
+  d3d_model_wall(draw_get_batch_stream(), x1, y1, z1, x2, y2, z2, hrep, vrep);
+  draw_batch_flush(batch_flush_immediate);
+}
+
+void d3d_draw_block(gs_scalar x1, gs_scalar y1, gs_scalar z1, gs_scalar x2, gs_scalar y2, gs_scalar z2, int texId, gs_scalar hrep, gs_scalar vrep, bool closed)
+{
+  draw_batch_begin_deferred(texId);
+  d3d_model_block(draw_get_batch_stream(), x1, y1, z1, x2, y2, z2, hrep, vrep, closed);
+  draw_batch_flush(batch_flush_immediate);
```

Comment:
Why are you flushing this each time? Doesn't that defeat the point of this batching? Or is that the point of this batch mode flag you are passing?

Reply:
That is the point of the `batch_flush_immediate` flag, the flush will return early if we're not in immediate batch mode. So in deferred mode, no, it is not flushed each time.